macro/src/lib.rs

```diff
@@ -85,61 +89,60 @@ pub fn subxt(args: TokenStream, input: TokenStream) -> TokenStream {
-    );
+    // Use the item module that the macro is on:
+    codegen.set_target_module(item_mod);
+
+    // Use the provided crate path:
+    if let Some(crate_path) = args.crate_path {
+        codegen.set_subxt_crate_path(crate_path)
+    }
 
-    for derives in &args.derive_for_type {
-        derives_registry.extend_for_type(
-            derives.path.clone(),
-            derives.derive.iter().cloned(),
-            vec![],
-        )
+    // Respect the boolean flags:
+    if args.runtime_types_only {
+        codegen.runtime_types_only();
     }
-    for attributes in &args.attributes_for_type {
-        derives_registry.extend_for_type(
-            attributes.path.clone(),
-            vec![],
-            attributes.attributes.iter().map(|a| a.0.clone()),
-        )
+    if args.no_default_derives {
+        codegen.disable_default_derives();
+    }
+    if args.no_default_substitutions {
+        codegen.disable_default_substitutes();
+    }
+    if !args.generate_docs.is_present() {
+        codegen.no_docs()
     }
 
-    let mut type_substitutes = if args.no_default_substitutions {
-        TypeSubstitutes::new()
-    } else {
-        TypeSubstitutes::with_default_substitutes(&crate_path)
-    };
-    let substitute_args_res: Result<(), _> = args.substitute_type.into_iter().try_for_each(|sub| {
-        sub.with
-            .try_into()
-            .and_then(|with| type_substitutes.insert(sub.path, with))
-    });
-
-    if let Err(err) = substitute_args_res {
-        return CodegenError::from(err).into_compile_error().into();
+    // Configure derives:
+    codegen.set_additional_global_derives(
+        args.derive_for_all_types
+            .unwrap_or_default()
+            .into_iter()
+            .collect(),
+    );
+    for d in args.derive_for_type {
+        codegen.add_derives_for_type(d.path, d.derive.into_iter());
     }
 
-    let should_gen_docs = args.generate_docs.is_present();
-    let unstable_metadata = args.unstable_metadata.is_present();
+    // Configure attributes:
+    codegen.set_additional_global_attributes(
+        args.attributes_for_all_types
+            .unwrap_or_default()
+            .into_iter()
+            .map(|a| a.0)
+            .collect(),
+    );
+    for d in args.attributes_for_type {
+        codegen.add_attributes_for_type(d.path, d.attributes.into_iter().map(|a| a.0))
+    }
+
+    // Insert type substitutions:
+    for sub in args.substitute_type.into_iter() {
+        codegen.set_type_substitute(sub.path, sub.with);
+    }
 
-    let runtime_api_generator = GenerateRuntimeApi::new(item_mod, crate_path)
-        .derives_registry(derives_registry)
-        .type_substitutes(type_substitutes)
-        .generate_docs(should_gen_docs)
-        .runtime_types_only(args.runtime_types_only);
+    // Do we want to fetch unstable metadata? This only works if fetching from a URL.
+    let unstable_metadata = args.unstable_metadata.is_present();
```

Comment:
What is the reasoning behind allowing unstable metadata only from URLs?

Reply:
Oh; this is just because if you provide the metadata, you can't pick what version it will be, but if you fetch from a URL you can decide what to fetch. No real need for this option at the mo since it wouldn't actually work, but when we start work on V16 metadata at some point it'll become useful again to try it out in Subxt :)